File: schedule_data/build_team_schedules.py
#!/usr/bin/env python3
"""
Build per-team Excel tabs from a Throw Down schedule CSV (group + playoff summary).

Usage:
  python3 schedule_data/build_team_schedules.py
  python3 schedule_data/build_team_schedules.py --csv path/to/schedule.csv --out path/to/out.xlsx
"""

from __future__ import annotations

import argparse
import csv
from collections import defaultdict
from dataclasses import dataclass
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter

HERE = Path(__file__).resolve().parent
DEFAULT_CSV = HERE / "schedule_may_27.csv"
DEFAULT_OUT = HERE / "The Throw Down_ 5th Edition Team Schedules - May 27.xlsx"
STREAM_COURT = "Court 1"

FILL_PLAY = PatternFill(start_color="C6EFCE", end_color="C6EFCE", fill_type="solid")
FILL_REF = PatternFill(start_color="FFF2CC", end_color="FFF2CC", fill_type="solid")
FILL_OFF = PatternFill(start_color="D9D9D9", end_color="D9D9D9", fill_type="solid")
FILL_HEADER = PatternFill(start_color="4472C4", end_color="4472C4", fill_type="solid")

FONT_TITLE = Font(bold=True, size=16)
FONT_SUBTITLE = Font(size=12)
FONT_TABLE_HEADER = Font(bold=True, color="FFFFFF")
ALIGN_WRAP = Alignment(wrap_text=True, vertical="top")


@dataclass(frozen=True)
class GameRow:
    date: str
    court: str
    phase: str
    group: str
    round_num: int
    home: str
    away: str
    referees: str

    @classmethod
    def from_dict(cls, row: dict[str, str]) -> GameRow:
        return cls(
            date=(row.get("Date") or "").strip(),
            court=(row.get("Court") or "").strip(),
            phase=(row.get("Phase") or "").strip(),
            group=(row.get("Group") or "").strip(),
            round_num=int((row.get("Round") or "0").strip()),
            home=_clean_team(row.get("Home Team", "")),
            away=_clean_team(row.get("Away Team", "")),
            referees=_clean_team(row.get("Referees", "")),
        )


def _clean_team(s: str) -> str:
    t = (s or "").strip()
    if t.lower() == "undefined":
        return ""
    return t


def load_games(path: Path) -> list[GameRow]:
    with path.open(newline="", encoding="utf-8-sig") as f:
        reader = csv.DictReader(f)
        return [GameRow.from_dict(dict(r)) for r in reader]


def group_phase_teams(games: list[GameRow]) -> list[str]:
    names: set[str] = set()
    for g in games:
        if g.phase != "Group Phase":
            continue
        if g.home:
            names.add(g.home)
        if g.away:
            names.add(g.away)
    return sorted(names)


def games_by_round(games: list[GameRow], phase: str) -> dict[int, list[GameRow]]:
    out: dict[int, list[GameRow]] = defaultdict(list)
    for g in games:
        if g.phase == phase:
            out[g.round_num].append(g)
    for r in out:
        out[r].sort(key=lambda x: (x.court, x.date))
    return dict(out)


def excel_sheet_title(name: str, used: set[str]) -> str:
    bad = '[]:*?/\\'
    s = "".join("_" if c in bad else c for c in name)
    if len(s) <= 31:
        base = s
    else:
        base = s[:31]
    candidate = base
    n = 2
    while candidate in used:
        suffix = f" ({n})"
        candidate = (base[: 31 - len(suffix)] + suffix) if len(base) + len(suffix) > 31 else base + suffix
        n += 1
    used.add(candidate)
    return candidate


def group_phase_round_stats(
    team: str, gp_by_round: dict[int, list[GameRow]]
) -> tuple[int, int, int, int]:
    """Returns (home_rounds, away_rounds, ref_rounds, off_rounds) for group phase rounds 1–10."""
    gp_rounds = sorted(k for k in gp_by_round if 1 <= k <= 10)
    home_c = away_c = ref_c = off_c = 0
    for rnd in gp_rounds:
        games = gp_by_round[rnd]
        if not games:
            off_c += 1
            continue
        if next((g for g in games if g.home == team), None):
            home_c += 1
        elif next((g for g in games if g.away == team), None):
            away_c += 1
        elif next((g for g in games if g.referees == team), None):
            ref_c += 1
        else:
            off_c += 1
    return home_c, away_c, ref_c, off_c


def group_phase_stream_play_rounds(
    team: str, gp_by_round: dict[int, list[GameRow]]
) -> list[int]:
    """Rounds (1–10) where the team is home or away on the stream court (Court 1)."""
    rounds: list[int] = []
    for rnd in sorted(k for k in gp_by_round if 1 <= k <= 10):
        for g in gp_by_round[rnd]:
            if g.court != STREAM_COURT:
                continue
            if g.home == team or g.away == team:
                rounds.append(rnd)
                break
    return rounds


def write_summary_sheet(
    ws,
    teams: list[str],
    gp_by_round: dict[int, list[GameRow]],
) -> None:
    ws["A1"] = "Group phase — team summary"
    ws["A1"].font = FONT_TITLE
    ws.merge_cells("A1:H1")

    ws["A2"] = "Throw Down — group rounds 1–10"
    ws["A2"].font = FONT_SUBTITLE
    ws.merge_cells("A2:H2")

    headers = [
        "Team",
        "Home",
        "Away",
        "Ref",
        "Off",
        "Total rounds",
        "Playing total",
        "Court 1 (stream) — playing rounds",
    ]
    hr = 4
    for col, h in enumerate(headers, start=1):
        cell = ws.cell(row=hr, column=col, value=h)
        cell.font = FONT_TABLE_HEADER
        cell.fill = FILL_HEADER
        cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)

    row_idx = hr + 1
    for team in teams:
        h, a, r, o = group_phase_round_stats(team, gp_by_round)
        total = h + a + r + o
        play = h + a
        c1_rounds = group_phase_stream_play_rounds(team, gp_by_round)
        c1_str = ", ".join(str(x) for x in c1_rounds) if c1_rounds else "—"
        values = [team, h, a, r, o, total, play, c1_str]
        for col, val in enumerate(values, start=1):
            c = ws.cell(row=row_idx, column=col, value=val)
            c.alignment = ALIGN_WRAP
            if col in (2, 3):
                c.fill = FILL_PLAY
            elif col == 4:
                c.fill = FILL_REF
            elif col == 5:
                c.fill = FILL_OFF
            elif col == 8:
                c.fill = FILL_PLAY
        row_idx += 1

    widths = [28, 10, 10, 10, 10, 14, 14, 32]
    for i, w in enumerate(widths, start=1):
        ws.column_dimensions[get_column_letter(i)].width = w


def write_team_sheet(
    ws,
    team: str,
    gp_by_round: dict[int, list[GameRow]],
    playoff_summary: list[tuple[int, str, str, str]],
) -> None:
    ws["A1"] = team
    ws["A1"].font = FONT_TITLE
    ws.merge_cells("A1:G1")

    ws["A2"] = "Throw Down — schedule export"
    ws["A2"].font = FONT_SUBTITLE
    ws.merge_cells("A2:G2")

    headers = ["Round", "Time", "Court", "Status", "Home Team", "Away Team", "Ref Team"]
    hr = 4
    for col, h in enumerate(headers, start=1):
        cell = ws.cell(row=hr, column=col, value=h)
        cell.font = FONT_TABLE_HEADER
        cell.fill = FILL_HEADER
        cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)

    row_idx = hr + 1
    gp_rounds = sorted(k for k in gp_by_round if 1 <= k <= 10)

    for rnd in gp_rounds:
        games = gp_by_round[rnd]
        if not games:
            continue
        round_time = games[0].date

        playing_home = next((g for g in games if g.home == team), None)
        playing_away = next((g for g in games if g.away == team), None)
        reffing = next((g for g in games if g.referees == team), None)

        if playing_home:
            status = "PLAYING (Home)"
            g = playing_home
            fill = FILL_PLAY
            home_v, away_v, ref_v = g.home, g.away, g.referees
            court_v = g.court
        elif playing_away:
            status = "PLAYING (Away)"
            g = playing_away
            fill = FILL_PLAY
            home_v, away_v, ref_v = g.home, g.away, g.referees
            court_v = g.court
        elif reffing:
            status = "REFFING"
            g = reffing
            fill = FILL_REF
            home_v, away_v, ref_v = g.home, g.away, g.referees
            court_v = g.court
        else:
            status = "OFF"
            fill = FILL_OFF
            home_v, away_v, ref_v = "—", "—", "—"
            court_v = "—"

        values = [rnd, round_time, court_v, status, home_v, away_v, ref_v]
        for col, val in enumerate(values, start=1):
            c = ws.cell(row=row_idx, column=col, value=val)
            c.fill = fill
            c.alignment = ALIGN_WRAP
        row_idx += 1

    n_play = sum(
        1
        for rnd in gp_rounds
        if next((g for g in gp_by_round[rnd] if g.home == team or g.away == team), None)
    )
    n_ref = sum(
        1
        for rnd in gp_rounds
        if next((g for g in gp_by_round[rnd] if g.referees == team), None)
    )
    n_off = len(gp_rounds) - n_play - n_ref
    row_idx += 1
    summary = (
        f"Group phase summary — Playing: {n_play} rounds • "
        f"Reffing: {n_ref} rounds • Off: {n_off} rounds"
    )
    csum = ws.cell(row=row_idx, column=1, value=summary)
    csum.font = Font(italic=True, size=11)
    ws.merge_cells(start_row=row_idx, start_column=1, end_row=row_idx, end_column=7)

    row_idx += 1
    ws.cell(row=row_idx, column=1, value="Playoffs — Bracket TBD Based on Seeding").font = Font(
        bold=True, size=12
    )
    ws.merge_cells(start_row=row_idx, start_column=1, end_row=row_idx, end_column=4)
    row_idx += 1

    ph = ["Round", "Time", "Phase / matches", "Courts"]
    for col, h in enumerate(ph, start=1):
        cell = ws.cell(row=row_idx, column=col, value=h)
        cell.font = FONT_TABLE_HEADER
        cell.fill = FILL_HEADER
        cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)
    row_idx += 1

    for rnd, tme, desc, courts in playoff_summary:
        for col, val in enumerate([rnd, tme, desc, courts], start=1):
            c = ws.cell(row=row_idx, column=col, value=val)
            c.alignment = ALIGN_WRAP
        row_idx += 1

    widths = [10, 22, 12, 18, 22, 22, 22]
    for i, w in enumerate(widths, start=1):
        ws.column_dimensions[get_column_letter(i)].width = w


def build_playoff_summary(playoff_by_round: dict[int, list[GameRow]]) -> list[tuple[int, str, str, str]]:
    summary: list[tuple[int, str, str, str]] = []
    for rnd in sorted(playoff_by_round):
        games = playoff_by_round[rnd]
        if not games:
            continue
        tme = games[0].date
        groups = [g.group for g in games]
        unique_groups = list(dict.fromkeys(groups))
        desc = " | ".join(unique_groups) if unique_groups else "Playoff block"
        courts_nums = []
        for g in games:
            c = g.court.replace("Court ", "").strip()
            if c.isdigit():
                courts_nums.append(int(c))
        if courts_nums:
            lo, hi = min(courts_nums), max(courts_nums)
            courts_str = f"Courts {lo}–{hi}" if lo != hi else f"Court {lo}"
        else:
            courts_str = ", ".join(g.court for g in games)
        summary.append((rnd, tme, desc, courts_str))
    return summary


def main() -> None:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--csv", type=Path, default=DEFAULT_CSV, help="Input schedule CSV")
    p.add_argument("--out", type=Path, default=DEFAULT_OUT, help="Output xlsx path")
    args = p.parse_args()

    games = load_games(args.csv)
    teams = group_phase_teams(games)
    gp_by_round = games_by_round(games, "Group Phase")
    playoff_by_round = games_by_round(games, "Playoff")
    playoff_summary = build_playoff_summary(playoff_by_round)

    wb = Workbook()
    default_sheet = wb.active
    wb.remove(default_sheet)

    ws_sum = wb.create_sheet("Summary", 0)
    write_summary_sheet(ws_sum, teams, gp_by_round)

    used_titles: set[str] = set()
    for team in teams:
        title = excel_sheet_title(team, used_titles)
        ws = wb.create_sheet(title)
        write_team_sheet(ws, team, gp_by_round, playoff_summary)

    args.out.parent.mkdir(parents=True, exist_ok=True)
    wb.save(args.out)
    print(f"Wrote summary + {len(teams)} team sheets to {args.out}")


if __name__ == "__main__":
    main()

File: app/lib/buildTeamSchedules.ts
import * as XLSX from 'xlsx';
import { splitCsvLine } from './scheduleParser';

export const STREAM_COURT = 'Court 1';

export interface GameRow {
  date: string;
  court: string;
  phase: string;
  group: string;
  roundNum: number;
  home: string;
  away: string;
  referees: string;
}

export interface PlayoffSummaryRow {
  round: number;
  time: string;
  description: string;
  courts: string;
}

function cleanTeam(value: string | undefined): string {
  const t = (value ?? '').trim();
  if (t.toLowerCase() === 'undefined') return '';
  return t;
}

function headerIndex(header: string[], name: string): number {
  const lower = name.toLowerCase();
  return header.findIndex((h) => h.includes(lower));
}

export function parseThrowdownScheduleCsv(csvText: string): GameRow[] {
  const normalized = csvText.replace(/^\uFEFF/, '');
  const lines = normalized.split('\n').filter((l) => l.trim());
  if (lines.length < 2) return [];

  const header = splitCsvLine(lines[0]).map((h) => h.trim().toLowerCase());
  const dateIdx = headerIndex(header, 'date');
  const courtIdx = headerIndex(header, 'court');
  const phaseIdx = headerIndex(header, 'phase');
  const groupIdx = headerIndex(header, 'group');
  const roundIdx = headerIndex(header, 'round');
  const homeIdx = headerIndex(header, 'home team');
  const awayIdx = headerIndex(header, 'away team');
  const refIdx = headerIndex(header, 'referee');

  const rows: GameRow[] = [];
  for (let i = 1; i < lines.length; i++) {
    const cells = splitCsvLine(lines[i]);
    const get = (n: number) => (n >= 0 ? cells[n]?.trim() ?? '' : '');
    rows.push({
      date: get(dateIdx),
      court: get(courtIdx),
      phase: get(phaseIdx),
      group: get(groupIdx),
      roundNum: parseInt(get(roundIdx) || '0', 10) || 0,
      home: cleanTeam(get(homeIdx)),
      away: cleanTeam(get(awayIdx)),
      referees: cleanTeam(get(refIdx)),
    });
  }
  return rows;
}

export function groupPhaseTeams(games: GameRow[]): string[] {
  const names = new Set<string>();
  for (const g of games) {
    if (g.phase !== 'Group Phase') continue;
    if (g.home) names.add(g.home);
    if (g.away) names.add(g.away);
  }
  return [...names].sort();
}

export function gamesByRound(games: GameRow[], phase: string): Map<number, GameRow[]> {
  const out = new Map<number, GameRow[]>();
  for (const g of games) {
    if (g.phase !== phase) continue;
    const list = out.get(g.roundNum) ?? [];
    list.push(g);
    out.set(g.roundNum, list);
  }
  for (const [round, list] of out) {
    list.sort((a, b) => a.court.localeCompare(b.court) || a.date.localeCompare(b.date));
    out.set(round, list);
  }
  return out;
}

export function excelSheetTitle(name: string, used: Set<string>): string {
  const bad = '[]:*?/\\';
  let base = [...name].map((c) => (bad.includes(c) ? '_' : c)).join('');
  if (base.length > 31) base = base.slice(0, 31);

  let candidate = base;
  let n = 2;
  while (used.has(candidate)) {
    const suffix = ` (${n})`;
    candidate =
      base.length + suffix.length > 31 ? base.slice(0, 31 - suffix.length) + suffix : base + suffix;
    n += 1;
  }
  used.add(candidate);
  return candidate;
}

export function groupPhaseRoundStats(
  team: string,
  gpByRound: Map<number, GameRow[]>
): [number, number, number, number] {
  const gpRounds = [...gpByRound.keys()].filter((k) => k >= 1 && k <= 10).sort((a, b) => a - b);
  let homeC = 0;
  let awayC = 0;
  let refC = 0;
  let offC = 0;

  for (const rnd of gpRounds) {
    const games = gpByRound.get(rnd)!;
    if (games.some((g) => g.home === team)) homeC += 1;
    else if (games.some((g) => g.away === team)) awayC += 1;
    else if (games.some((g) => g.referees === team)) refC += 1;
    else offC += 1;
  }
  return [homeC, awayC, refC, offC];
}

export function groupPhaseStreamPlayRounds(
  team: string,
  gpByRound: Map<number, GameRow[]>
): number[] {
  const rounds: number[] = [];
  for (const rnd of [...gpByRound.keys()].filter((k) => k >= 1 && k <= 10).sort((a, b) => a - b)) {
    for (const g of gpByRound.get(rnd) ?? []) {
      if (g.court !== STREAM_COURT) continue;
      if (g.home === team || g.away === team) {
        rounds.push(rnd);
        break;
      }
    }
  }
  return rounds;
}

export function buildPlayoffSummary(
  playoffByRound: Map<number, GameRow[]>
): PlayoffSummaryRow[] {
  const summary: PlayoffSummaryRow[] = [];
  for (const rnd of [...playoffByRound.keys()].sort((a, b) => a - b)) {
    const games = playoffByRound.get(rnd) ?? [];
    if (games.length === 0) continue;

    const time = games[0].date;
    const uniqueGroups = [...new Set(games.map((g) => g.group))];
    const description = uniqueGroups.length > 0 ? uniqueGroups.join(' | ') : 'Playoff block';

    const courtNums = games
      .map((g) => g.court.replace('Court ', '').trim())
      .filter((c) => /^\d+$/.test(c))
      .map((c) => parseInt(c, 10));

    let courts: string;
    if (courtNums.length > 0) {
      const lo = Math.min(...courtNums);
      const hi = Math.max(...courtNums);
      courts = lo !== hi ? `Courts ${lo}–${hi}` : `Court ${lo}`;
    } else {
      courts = games.map((g) => g.court).join(', ');
    }

    summary.push({ round: rnd, time, description, courts });
  }
  return summary;
}

function buildSummarySheetRows(teams: string[], gpByRound: Map<number, GameRow[]>): string[][] {
  const rows: string[][] = [
    ['Group phase — team summary'],
    ['Throw Down — group rounds 1–10'],
    [],
    [
      'Team',
      'Home',
      'Away',
      'Ref',
      'Off',
      'Total rounds',
      'Playing total',
      'Court 1 (stream) — playing rounds',
    ],
  ];

  for (const team of teams) {
    const [h, a, r, o] = groupPhaseRoundStats(team, gpByRound);
    const total = h + a + r + o;
    const play = h + a;
    const c1Rounds = groupPhaseStreamPlayRounds(team, gpByRound);
    const c1Str = c1Rounds.length > 0 ? c1Rounds.join(', ') : '—';
    rows.push([team, String(h), String(a), String(r), String(o), String(total), String(play), c1Str]);
  }
  return rows;
}

function buildTeamSheetRows(
  team: string,
  gpByRound: Map<number, GameRow[]>,
  playoffSummary: PlayoffSummaryRow[]
): string[][] {
  const rows: string[][] = [
    [team],
    ['Throw Down — schedule export'],
    [],
    ['Round', 'Time', 'Court', 'Status', 'Home Team', 'Away Team', 'Ref Team'],
  ];

  const gpRounds = [...gpByRound.keys()].filter((k) => k >= 1 && k <= 10).sort((a, b) => a - b);
  let nPlay = 0;
  let nRef = 0;

  for (const rnd of gpRounds) {
    const games = gpByRound.get(rnd)!;
    const roundTime = games[0].date;
    const playingHome = games.find((g) => g.home === team);
    const playingAway = games.find((g) => g.away === team);
    const reffing = games.find((g) => g.referees === team);

    if (playingHome) {
      nPlay += 1;
      rows.push([
        String(rnd),
        roundTime,
        playingHome.court,
        'PLAYING (Home)',
        playingHome.home,
        playingHome.away,
        playingHome.referees,
      ]);
    } else if (playingAway) {
      nPlay += 1;
      rows.push([
        String(rnd),
        roundTime,
        playingAway.court,
        'PLAYING (Away)',
        playingAway.home,
        playingAway.away,
        playingAway.referees,
      ]);
    } else if (reffing) {
      nRef += 1;
      rows.push([
        String(rnd),
        roundTime,
        reffing.court,
        'REFFING',
        reffing.home,
        reffing.away,
        reffing.referees,
      ]);
    } else {
      rows.push([String(rnd), roundTime, '—', 'OFF', '—', '—', '—']);
    }
  }

  const nOff = gpRounds.length - nPlay - nRef;
  rows.push([]);
  rows.push([
    `Group phase summary — Playing: ${nPlay} rounds • Reffing: ${nRef} rounds • Off: ${nOff} rounds`,
  ]);
  rows.push([]);
  rows.push(['Playoffs — Bracket TBD Based on Seeding']);
  rows.push(['Round', 'Time', 'Phase / matches', 'Courts']);

  for (const row of playoffSummary) {
    rows.push([String(row.round), row.time, row.description, row.courts]);
  }

  return rows;
}

function applyColumnWidths(ws: XLSX.WorkSheet, widths: number[]): void {
  ws['!cols'] = widths.map((wch) => ({ wch }));
}

export function buildTeamSchedulesWorkbook(csvText: string): XLSX.WorkBook {
  const games = parseThrowdownScheduleCsv(csvText);
  const teams = groupPhaseTeams(games);
  const gpByRound = gamesByRound(games, 'Group Phase');
  const playoffByRound = gamesByRound(games, 'Playoff');
  const playoffSummary = buildPlayoffSummary(playoffByRound);

  const workbook = XLSX.utils.book_new();

  const summarySheet = XLSX.utils.aoa_to_sheet(buildSummarySheetRows(teams, gpByRound));
  applyColumnWidths(summarySheet, [28, 10, 10, 10, 10, 14, 14, 32]);
  XLSX.utils.book_append_sheet(workbook, summarySheet, 'Summary');

  const usedTitles = new Set<string>(['Summary']);
  for (const team of teams) {
    const title = excelSheetTitle(team, usedTitles);
    const teamSheet = XLSX.utils.aoa_to_sheet(
      buildTeamSheetRows(team, gpByRound, playoffSummary)
    );
    applyColumnWidths(teamSheet, [10, 22, 12, 18, 22, 22, 22]);
    XLSX.utils.book_append_sheet(workbook, teamSheet, title);
  }

  return workbook;
}

export function buildTeamSchedulesBuffer(csvText: string): Buffer {
  const workbook = buildTeamSchedulesWorkbook(csvText);
  return XLSX.write(workbook, { type: 'buffer', bookType: 'xlsx' }) as Buffer;
}

export function defaultTeamSchedulesFilename(): string {
  return 'Throw_Down_Team_Schedules.xlsx';
}
